api/grpc/controller_test.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2021 Alvalor S.A.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package grpc
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	tmock "github.com/stretchr/testify/mock"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/models/dps/mock"
+)
+
+func TestNewController(t *testing.T) {
+	c := NewController(nil)
+	assert.NotNil(t, c)
+}
+
+func TestController_GetRegister(t *testing.T) {
+	var (
+		testHeight uint64 = 128
+		lastHeight uint64 = 256
+
+		testKey   = []byte(`testKey`)
+		testValue = []byte(`testValue`)
+	)
+
+	tests := []struct {
+		desc string
+
+		reqHeight *uint64
+		reqKey    []byte
+
+		mockValue []byte
+		mockErr   error
+
+		wantHeight uint64
+		wantResp   *GetRegisterResponse
+		wantErr    assert.ErrorAssertionFunc
+	}{
+		{
+			desc: "nominal case, height given",
+
+			reqHeight: &testHeight,
+			reqKey:    testKey,
+
+			mockValue: testValue,
+
+			wantHeight: testHeight,
+			wantResp: &GetRegisterResponse{
+				Height: testHeight,
+				Key:    testKey,
+				Value:  testValue,
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			desc: "nominal case, no height given",
+
+			reqKey: testKey,
+
+			mockValue: testValue,
+
+			wantHeight: lastHeight,
+			wantResp: &GetRegisterResponse{
+				Height: lastHeight,
+				Key:    testKey,
+				Value:  testValue,
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			desc: "state error",
+
+			reqKey: testKey,
+
+			mockErr: errors.New("dummy error"),
+
+			wantHeight: lastHeight,
+			wantResp:   nil,
+			wantErr:    assert.Error,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			t.Parallel()
+
+			m := mock.NewState()
+			m.LastState.On("Height").Return(lastHeight).Once()
+			m.RawState.On("WithHeight", test.wantHeight).Return(m.RawState).Once()
+			m.RawState.On("Get", test.reqKey).Return(test.mockValue, test.mockErr)
+
+			c := &Controller{
+				state: m,
+			}
+
+			req := &GetRegisterRequest{
+				Height: test.reqHeight,
+				Key:    test.reqKey,
+			}
+
+			got, err := c.GetRegister(context.Background(), req)
+			test.wantErr(t, err)
+
+			if test.wantResp != nil {
+				assert.Equal(t, test.wantResp, got)
+			}
+
+			m.AssertExpectations(t)
+		})
+	}
+}
+
+func TestController_GetValues(t *testing.T) {
```

Comment:
In terms of testing theory, it makes sense to have one test case for each border value. In this case, this means we should have one test for when we have zero keys (that would reveal that we should actually decide how to handle the case!), and one with more than one key; two would be fine, but I usually do three for those cases. There are some edge cases where functions would work correctly for two values and mess up the order with three.